File: input/reference_haversine.js
const EARTH_RADIUS = 6372.8;

function square(x) {
  return x * x;
}

function degToRad(degrees) {
  return degrees * Math.PI / 180;
}

function haversine(x0, y0, x1, y1) {
  const lon0 = degToRad(x0);
  const lat0 = degToRad(y0);
  const lon1 = degToRad(x1);
  const lat1 = degToRad(y1);

  const dLon = lon1 - lon0;
  const dLat = lat1 - lat0;

  const a =
    square(Math.sin(dLat / 2)) +
    Math.cos(lat0) * Math.cos(lat1) * square(Math.sin(dLon / 2));

  const c = 2 * Math.asin(Math.sqrt(a));

  return EARTH_RADIUS * c;
}

module.exports = {
  haversine,
};

File: reference_haversine.c
#include "reference_haversine.h"

#define EARTH_RADIUS 6372.8
#define PI 3.14159265358979323846

static double square(double x){
  return x*x;
}

static double deg_to_rad(double deg){
  double rad = deg * PI / 180.0;
  return rad;
}

double reference_haversine(pair_t pairs){

  double lon0 = deg_to_rad(pairs.x0);
  double lat0 = deg_to_rad(pairs.y0);
  double lon1 = deg_to_rad(pairs.x1);
  double lat1 = deg_to_rad(pairs.y1);

  double diff_lon = lon1 - lon0;
  double diff_lat = lat1 - lat0;


  double a =
    square(sin(diff_lat / 2)) + cos(lat0) * cos(lat1) * square(sin(diff_lon / 2));

  double c = 2 * asin(sqrt(a));

  return EARTH_RADIUS * c;
}
